module/zfs/arc.c

```diff
@@ -4092,13 +4150,53 @@ arc_evict_state(arc_state_t *state, arc_buf_contents_t type, uint64_t spa,
 		multilist_sublist_unlock(mls);
 	}
 
+	if (use_evcttq) {
+		evarg = kmem_alloc(sizeof (*evarg) * nthreads, KM_NOSLEEP);
+		/*
+		 * Fall back to the regular single evict if it is not possible
+		 * to allocate memory for the task queue entries.
+		 */
+		if (evarg == NULL)
+			use_evcttq = B_FALSE;
+	}
+
 	/*
 	 * While we haven't hit our target number of bytes to evict, or
 	 * we're evicting all available buffers.
 	 */
 	while (total_evicted < bytes) {
 		int sublist_idx = multilist_get_random_index(ml);
 		uint64_t scan_evicted = 0;
+		uint64_t left = (bytes == ARC_EVICT_ALL ? bytes :
+		    bytes - total_evicted);
+		uint64_t evict;
+		uint_t ntasks;
+		uint_t m = 0;
+
+		if (use_evcttq) {
+			/*
+			 * Scaling algorithm:
+			 * 1. Divide the left bytes by chunks each
+			 *    MIN_EVICT_SIZE.
+			 * 2. Equaly distribute the chunks between the threads.
+			 * 3. First (m) threads get an additional chunk each.
+			 */
+			if (left == ARC_EVICT_ALL) {
+				ntasks = nthreads;
+				evict = left;
+			} else if (left > nthreads * MIN_EVICT_SIZE) {
+				ntasks = nthreads;
+				uint64_t n = DIV_ROUND_UP(left, MIN_EVICT_SIZE);
+				evict = (n / nthreads) * MIN_EVICT_SIZE;
+				m = n % nthreads;	 /* additional chunks */
```

Comment:
You are again start aligning on `MIN_EVICT_SIZE`, which I several times repeated should not be needed and just complicates the code.